File: demo-task-executor-scala/src/test/scala/demo/executor/AbstractAsyncExecutorTest.scala
package demo.executor

import java.util.concurrent.TimeUnit

import demo.CallableHelper._
import org.testng.Assert._

/**
  * Abstract unit test for {[demo.executor.AsyncExecutor]}.
  */
abstract class AbstractAsyncExecutorTest extends AbstractConcurrencyTest {

    protected val TASK_DURATION_MSEC: Long = 1000

    protected def getTestTask(value: Integer, action: (Option[Integer], Option[Exception]) => Unit): Task[Integer] = {
        getTestTask(value, classOf[Integer], action)
    }

    protected def getTestTaskWithResult(value: Integer): Task[Integer] = {
        getTestTaskWithResult(value, classOf[Integer])
    }

    protected def getTestTaskWithException(value: Integer): Task[Integer] = {
        getTestTaskWithException(value, classOf[Integer])
    }

    protected def getTestTask(consumer: java.util.List[Integer], value: Integer): Task[Integer] = {
        getTestTask(value, { (result: Option[Integer], exception: Option[Exception]) =>
            if (exception.nonEmpty) {
                fail()
            } else {
                consumer.add(value)
            }
        })
    }

    private def getTestTask[T](value: T, clazz: Class[T], action: (Option[T], Option[Exception]) => Unit): Task[T] = {
        new Task("task " + value, clazz,
            callable {
                () => {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC)
                    value
                }
            },
            action)
    }

    private def getTestTaskWithResult[T](value: T, clazz: Class[T]): Task[T] = {
        new Task("task " + value, clazz,
            callable {
                () => {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC)
                    value
                }
            }, { (result: Option[T], exception: Option[Exception]) =>
                if (exception.nonEmpty) {
                    fail()
                } else {
                    assertEquals(result.get, value)
                }
            })
    }

    private def getTestTaskWithException[T](value: T, clazz: Class[T]): Task[T] = {
        new Task[T]("task " + value, clazz,
            callable {
                () => {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC)
                    throw new IllegalArgumentException()
                }
            }, { (result: Option[T], exception: Option[Exception]) =>
                if (exception.nonEmpty) {
                    assertTrue(exception.get.isInstanceOf[IllegalArgumentException])
                } else {
                    fail()
                }
            })
    }
}

File: demo-task-executor-java/src/test/java/demo/executor/AbstractAsyncExecutorTest.java
package demo.executor;

import java.util.List;
import java.util.concurrent.TimeUnit;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;
import static org.testng.Assert.fail;

/**
 * Abstract unit test for {{@link demo.executor.AsyncExecutor}}.
 */
abstract public class AbstractAsyncExecutorTest extends AbstractConcurrencyTest {

    protected static final int TASK_DURATION_MSEC = 1000;

    protected Task<Integer> getTestTask(int value, Action<Integer> action) {
        return getTestTask(value, Integer.class, action);
    }

    protected Task<Integer> getTestTaskWithResult(int value) {
        return getTestTaskWithResult(value, Integer.class);
    }

    protected Task<Integer> getTestTaskWithException(int value) {
        return getTestTaskWithException(value, Integer.class);
    }

    protected Task getTestTask(List<Integer> consumer, int value) {
        return getTestTask(value,
                (result, exception) -> {
                    if (exception.isPresent()) {
                        fail();
                    } else {
                        consumer.add(value);
                    }
                }
        );
    }

    private <T> Task<T> getTestTask(T value, Class<T> clazz, Action<T> action) {
        return new Task<>("task " + value, clazz,
                () -> {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC);
                    return value;
                },
                action
        );
    }

    private <T> Task<T> getTestTaskWithResult(T value, Class<T> clazz) {
        return new Task<>("task " + value, clazz,
                () -> {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC);
                    return value;
                },
                (result, exception) -> {
                    if (exception.isPresent()) {
                        fail();
                    } else {
                        assertEquals(result.get(), value);
                    }
                }
        );
    }

    private <T> Task<T> getTestTaskWithException(T value, Class<T> clazz) {
        return new Task<>("task " + value, clazz,
                () -> {
                    TimeUnit.MILLISECONDS.sleep(TASK_DURATION_MSEC);
                    throw new IllegalArgumentException();
                },
                (result, exception) -> {
                    if (exception.isPresent()) {
                        assertTrue(exception.get() instanceof IllegalArgumentException);
                    } else {
                        fail();
                    }
                }
        );
    }
}
